File: backend/app/services/payroll.py
def calculate_payroll(user, hours_summary: dict) -> dict:
    """Calculate payroll from user settings and hours summary"""
    hourly_wage = user.hourly_wage

    base_salary = round(hours_summary["regular_hours"] * hourly_wage)
    overtime_pay = round(
        hours_summary["overtime_hours"] * hourly_wage * user.overtime_rate
    )
    late_night_pay = round(
        hours_summary["late_night_hours"] * hourly_wage * (user.late_night_rate - 1.0)
    )
    holiday_pay = round(
        hours_summary["holiday_hours"] * hourly_wage * user.holiday_rate
    )

    gross_pay = (
        base_salary
        + overtime_pay
        + late_night_pay
        + holiday_pay
        + user.transportation
        + user.fixed_allowance
    )

    return {
        "base_salary": base_salary,
        "overtime_pay": overtime_pay,
        "late_night_pay": late_night_pay,
        "holiday_pay": holiday_pay,
        "transportation": user.transportation,
        "allowances": user.fixed_allowance,
        "deductions": 0.0,
        "gross_pay": gross_pay,
    }

File: backend/src/services/payroll.js
'use strict';

function calculatePayroll(user, hours) {
  const w = user.hourly_wage || 0;
  const base = Math.round(hours.regular_hours * w);
  const overtime = Math.round(hours.overtime_hours * w * (user.overtime_rate || 1.25));
  const lateNight = Math.round(hours.late_night_hours * w * ((user.late_night_rate || 1.25) - 1));
  const holiday = Math.round(hours.holiday_hours * w * (user.holiday_rate || 1.35));
  const gross = base + overtime + lateNight + holiday + (user.transportation || 0) + (user.fixed_allowance || 0);
  return {
    base_salary: base,
    overtime_pay: overtime,
    late_night_pay: lateNight,
    holiday_pay: holiday,
    transportation: user.transportation || 0,
    allowances: user.fixed_allowance || 0,
    deductions: 0,
    gross_pay: gross,
  };
}

module.exports = { calculatePayroll };
